---
title: "Preprocessing of Hammond et al. Lifespan Microglia"
author: "James Choi"
date: "Last compiled:`r Sys.Date()`"
output: 
  html_document
editor_options:
  chunk_output_type: 
    console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning=FALSE, message=FALSE, fig.align='center', 
                      tidy.opts=list(width.cutoff=80), tidy=TRUE,
                      results='hold')
```

```{r}
if(!grepl('scripts', getwd())) {setwd('scripts')}
results_out <- '../results/GSE121654_Hammond/'
data_out <- '../data/GSE121654_Hammond/'
dir.create(path = data_out)
```


GEO Accession: GSE121654

Data was provided by the following:

Title: Complex cell-state changes revealed by single cell RNA sequencing of 76,149 microglia throughout the mouse lifespan and in the injured brain

Citation: Hammond TR, Dufort C, Dissing-Olesen L, Giera S et al. Single-Cell RNA Sequencing of Microglia throughout the Mouse Lifespan and in the Injured Brain Reveals Complex Cell-State Changes. Immunity 2019 Jan 15;50(1):253-271.e6. PMID: 30471926

[Link to paper](DOI: 10.1016/j.immuni.2018.11.004)


```{r download_decompress}
dlcheck <- download.file(
  url = 'https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE121654&format=file',
  destfile = paste0(data_out, 'GSE121654_RAW.tar'),
  mode = 'wb' # the TAR file is a binary
)
if (dlcheck != 0) {
  stop('GSE121654 download unsuccessful')
} else {
  untar(tarfile = paste0(data_out, 'GSE121654_RAW.tar'), 
        exdir = paste0(data_out, 'GSE121654_RAW/'))
}
```


From GEO:

>Overall design: 41 total animals, 3-4 replicates per timepoint and condition. E14.5, P4/P5, P30, P100, P540, and Injury


Group-sample name key:

* P: post-natal
* E: embryonic
* M: male
* F: female
* Percoll vs no percoll, LPC vs saline


```{r import_wt_data}
tmp_keep <- c('E14','P4','P5','P30','P100','Old')
tmp_keep <- paste(tmp_keep, collapse = '|')
tmp_remove <- c('percoll', 'nopercoll', 'LPC', 'SALINE')
tmp_remove <- paste(tmp_remove, collapse = '|')

sample_files <- list.files(
  path = paste0(data_out, 'GSE121654_RAW/'), 
  full.names = TRUE
)
sample_files <- sample_files[grepl(tmp_keep, sample_files) & 
                               !grepl(tmp_remove, sample_files)]
# These files were part of Percoll experiment (I think). Because it's unclear, 
# they are removed.
sample_files <- sample_files[-c(33,34,35)]

counts <- vector(mode = 'list', length = length(sample_files))
for (i in 1:length(sample_files)) {
  message('Importing:', sample_files[i])
  tmp_counts <- read.table(
    file = gzfile(sample_files[i]),
    sep = '\t',
    header = TRUE,
    row.names = 1
  )
  message('nrow:', nrow(tmp_counts), '|ncol:', ncol(tmp_counts))
  tmp_counts <- as.matrix(tmp_counts)
  counts[[i]] <- Matrix::Matrix(data = tmp_counts, sparse = TRUE)
}
rm(list = ls()[grep('tmp', ls())])
```

Set unique cell barcodes to track back to original study sample.

```{r unique_barcodes}
sample_text <- strsplit(sample_files, '/')
sample_text <- sapply(sample_text, FUN = function(x) x[length(x)])
sample_text <- sapply(strsplit(sample_text, '\\.'), `[`, 1)
sample_names <- sapply(strsplit(sample_text, '_'), `[`, 1)

if (length(sample_names) != length(counts)) {
  stop('Number of sample names not equal to number of count matrices')
}
counts <- lapply(
  X = counts,
  FUN = function(x) {
    colnames(x) <- sapply(strsplit(x = colnames(x), split =  '\\.'), `[`, 1)
    return(x)
  }
)
for (i in 1:length(counts)) {
  colnames(counts[[i]]) <- paste(
    colnames(counts[[i]]),
    sample_names[i],
    sep = '_'
  )
}
```

Merge count matrices. 

```{r merge_counts}
genes <- lapply(X = counts, FUN = rownames)
universe <- Reduce(f = union, x = genes)
for (i in 1:length(counts)) {
  tmp_missing_genes <- universe[which(!universe %in% rownames(counts[[i]]))]
  tmp_missing_counts <- Matrix::Matrix(
    data = 0,
    nrow = length(tmp_missing_genes),
    ncol = ncol(counts[[i]]),
    sparse = TRUE
  )
  rownames(tmp_missing_counts) <- tmp_missing_genes
  colnames(tmp_missing_counts) <- colnames(counts[[i]])
  counts[[i]] <- rbind(counts[[i]], tmp_missing_counts)
  counts[[i]] <- counts[[i]][universe,]
}
sapply(counts, dim)
counts <- Reduce(f = cbind, x = counts)

# if (!all(sapply(genes[2:length(genes)], identical, genes[[1]]))) {
#   shared_genes <- Reduce(f = intersect, x = genes)
#   counts <- lapply(X = counts, FUN = function(x) x[shared_genes,])
#   counts <- Reduce(f = cbind, x = counts)
# } else {
#   counts <- Reduce(f = cbind, x = counts)
# }
# dim(counts)
```

Setup metadata including experimental details.

```{r metadata}
# Set up metadata and experimental details
metadata <- data.frame(
  'study' = 'Hammond2018',
  'sample_id' = sapply(strsplit(colnames(counts), '_'), `[`, 2),
  'tissue' = 'whole_brain',
  'injury_model' = 'Healthy'
)
rownames(metadata) <- colnames(counts)
# cells per sample
tmp_count <- sapply(strsplit(colnames(counts), '_'), `[`, 2)
tmp_count <- c(table(tmp_count))
# age
tmp_age <- sapply(strsplit(sample_text, '_'), `[`, 2)
tmp_age <- plyr::mapvalues(x = tmp_age, from = 'Old', to = 'P540')
tmp_age <- paste(tmp_age, 'Ctrl', sep = '.')
tmp_age <- rep(tmp_age, times = tmp_count)
metadata$injury <- tmp_age
# sex
tmp_sex <- sapply(strsplit(sample_text, '_'), `[`, 3)
tmp_sex <- toupper(substr(tmp_sex, 1, 1))
tmp_sex <- rep(tmp_sex, times = tmp_count)
metadata$sex <- tmp_sex
```

Save

```{r save}
Hammond2018 <- list(counts = counts, metadata = metadata)
saveRDS(Hammond2018, file = paste0(data_out, 'Hammond2018.rds'))
```

```{r}
sessionInfo()
```

